jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java

```diff
@@ -581,29 +588,69 @@ private Pool.Entry<RetainableByteBuffer> evict()
         }
     }
 
-    private static class Buffer extends AbstractRetainableByteBuffer
+    private interface Removable
+    {
+        boolean remove();
+    }
+
+    private class ReservedBuffer extends AbstractRetainableByteBuffer implements Removable
     {
-        private final Consumer<RetainableByteBuffer> _releaser;
+        private final RetainedBucket _bucket;
+        private final AtomicBoolean _removed = new AtomicBoolean();
+
+        private ReservedBuffer(ByteBuffer buffer, RetainedBucket bucket)
+        {
+            super(buffer);
+            _bucket = Objects.requireNonNull(bucket);
+        }
+
+        @Override
+        public boolean release()
+        {
+            boolean released = super.release();
+            if (released && _removed.compareAndSet(false, true))
+                reserve(_bucket, getByteBuffer());
+            return released;
+        }
+
+        @Override
+        public boolean remove()
+        {
+            super.release();
+            return _removed.compareAndSet(false, true);
+        }
+    }
+
+    private class Buffer extends AbstractRetainableByteBuffer implements Removable
+    {
+        private final RetainedBucket _bucket;
+        private final Pool.Entry<RetainableByteBuffer> _entry;
         private int _usages;
 
-        private Buffer(ByteBuffer buffer, Consumer<RetainableByteBuffer> releaser)
+        private Buffer(ByteBuffer buffer, RetainedBucket bucket, Pool.Entry<RetainableByteBuffer> entry)
         {
             super(buffer);
-            this._releaser = releaser;
+            _bucket = Objects.requireNonNull(bucket);
+            _entry = Objects.requireNonNull(entry);
         }
 
         @Override
         public boolean release()
         {
             boolean released = super.release();
             if (released)
-            {
-                if (_releaser != null)
-                    _releaser.accept(this);
-            }
+                ArrayByteBufferPool.this.release(_bucket, _entry);
             return released;
         }
 
+        @Override
+        public boolean remove()
+        {
+            boolean removed = ArrayByteBufferPool.this.remove(_bucket, _entry);
+            release();
```

Comment:
Calling release here looks wrong: this would reset the buffer - which we do not want here - and increment the released stat and the removed stat which would be incremented twice because `ArrayByteBufferPool.remove()` already incremented it once.

But calling `super.release()` is still needed to decrement the ref counter.